### ISSUE TITLE
BackgroundService blocked the execution of whole host

### ISSUE THREAD

**Issue by vova-lantsov-dev** [aspnet/AspNetCore#12907]
### Describe the bug
When I run the specific `foreach` cycle in `BackgroundService`-derived class, it blocks the whole host from starting. Even second hosted service doesn't start. When I comment `foreach` cycle, everything works as expected.

### To Reproduce
TargetFramework: netcoreapp2.1
Version: 2.1.12

Use following hosted service: [NotificationRunner.cs](https://gist.github.com/vova-lantsov-dev/7b3b820df1c7a6b86dac6bb0593c8155)
And singleton: [NotificationService.cs](https://gist.github.com/vova-lantsov-dev/788c5816a137a81e0c2af5f4655b21bc)

### Expected behavior
BackgroundService.ExecuteAsync should work in background without blocking even if it has blocking code. As you can see in [NotificationService](https://gist.github.com/vova-lantsov-dev/788c5816a137a81e0c2af5f4655b21bc) class, cancellation of enumerator is based on IApplicationLifetime.ApplicationStopping, but anyway it shouldn't affect the host startup because BackgroundService is expected to run in background :)

### Screenshots
When `foreach` cycle exists
![image](https://user-images.githubusercontent.com/45858087/62523649-ecafbf00-b83c-11e9-855e-e0ba541d0c3e.png)
Then execution is blocked on this cycle
![image](https://user-images.githubusercontent.com/45858087/62525002-4d3ffb80-b83f-11e9-8ce2-713dda7e79ad.png)

But when `foreach` cycle is commented
![image](https://user-images.githubusercontent.com/45858087/62525057-647ee900-b83f-11e9-84d0-47f896735b98.png)
Then execution continues as expected
![image](https://user-images.githubusercontent.com/45858087/62525220-ae67cf00-b83f-11e9-83a9-c7ca1c45b82c.png)
(But why twice?)

### Additional context
```
.NET Core SDK (reflecting any global.json):
 Version:   3.0.100-preview7-012821
 Commit:    6348f1068a

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.17763
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\3.0.100-preview7-012821\

Host (useful for support):
  Version: 3.0.0-preview7-27912-14
  Commit:  4da6ee6450

.NET Core SDKs installed:
  2.1.700 [C:\Program Files\dotnet\sdk]
  2.1.701 [C:\Program Files\dotnet\sdk]
  2.1.801 [C:\Program Files\dotnet\sdk]
  2.2.300 [C:\Program Files\dotnet\sdk]
  2.2.301 [C:\Program Files\dotnet\sdk]
  2.2.401 [C:\Program Files\dotnet\sdk]
  3.0.100-preview7-012821 [C:\Program Files\dotnet\sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.11 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.12 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.2.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.2.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.11 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.12 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.2.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.2.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.0.0-preview7.19365.7 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.11 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.12 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.6 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.0.0-preview7-27912-14 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.0.0-preview7-27912-14 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
```

**Comment by vova-lantsov-dev** [comment]
Fixed by adding `await Task.Yield()` at the top of method

**Comment by DAllanCarr** [comment]
Yeah the method shouldn't be blocking. Wrapping in a `Task,Run` will also dispatch.

**Comment by rynowak** [comment]
Reopening this so we can look at it. I'm glad that you solved your problem but we should try to solve this so others don't hit it.

**Comment by rynowak** [comment]
@anurse @Tratcher - thoughts? 

It looks like if your run a bunch of CPU-bound code in `ExecuteAsync` then it will stall startup until that work reaches the first actual async point.

See: https://github.com/aspnet/Extensions/blob/master/src/Hosting/Abstractions/src/BackgroundService.cs#L33

In health checks I ended up inserting a `Task.Yield` to avoid this. We'd get an additional state machine in this case, 1 per service, but it seems like a worthwhile tradeoff.

**Comment by vova-lantsov-dev** [comment]
Async method executes synchronously till first `await` is reached (it means `BackgroundService.ExecuteAsync` won't return till `await` is called). So this problem is not related to ASP.NET Core, but to async compilation.

**Comment by vova-lantsov-dev** [comment]
But it would be great if you can solve this case

**Comment by Tratcher** [comment]
It's a trade off. There may be reasonable initialization code the service needs to do before letting the app continue. It's within the service's control how long it blocks.

**Comment by kostya9** [comment]
I have also encountered this issue and solved it in the same way as @vova-lantsov-dev did. However, my main concern is that this behavior is not straightforward, would be nice if this was more predictable or documented somewhere

**Comment by anurse** [comment]
I think it's reasonable to dispatch `ExecuteAsync` to the thread-pool in `StartAsync`. If the user wants to perform initialization that **must complete** before letting the app continue, they can override `StartAsync` itself, right?

```csharp
public class MyService: BackgroundService
{
	public override async Task StartAsync()
	{
		await InitializeAsync();
		await base.StartAsync();
	}

	public override async Task ExecuteAsync()
	{
		// Do background stuff...
	}
}
```

It's called **Background**Service. It seems odd that you can block the foreground in the default case. If we're not happy with overriding `StartAsync` we could add a new `InitializeAsync` that is called **and `await`ed** during `StartAsync`.

This would be a breaking change however, since existing services may depend on the initial synchronous work blocking the startup process.

**Comment by davidfowl** [comment]
Sorry I never did this change but enough people have hit it now I think we should just do it. https://github.com/aspnet/Extensions/tree/davidfowl/background-service
